Migrate to ESLint 9 and flat configs

philippfromme opened the pull request:
### Proposed Changes

* Migrates this plugin to ESLint 9, ESM, and [flat configs](https://eslint.org/docs/latest/use/migrate-to-9.0.0#-new-default-config-format-eslintconfigjs). 
* Adds tests for when there are no errors _and_ when there are errors (verifying that the configs actually work).
* Improves test organization

Closes https://github.com/bpmn-io/eslint-plugin-bpmn-io/issues/19

<!--

Add relevant context (issue fixed or related to), 
a capture of the UI changes (if any) as well as 
steps to try out your changes.

--> 

### Checklist

To ensure you provided everything we need to look at your PR:

* [x] **Brief textual description** of the changes present
* [ ] **Visual demo** attached
* [ ] **Steps to try out** present, i.e. [using the `@bpmn-io/sr` tool](https://github.com/bpmn-io/sr)
* [x] Related issue linked via `Closes {LINK_TO_ISSUE}` or `Related to {LINK_TO_ISSUE}`

<!--

Thanks for creating this pull request! ❤️

-->

nikku commented on configs/mocha.js:
```diff
@@ -1,15 +1,12 @@
-  rules: {
-    'mocha/no-exclusive-tests': 2
-  },
-  globals: {
```
This is a breaking change that will prevent us from using `expect` in environments such as diagram-js where it is provided externally.

nikku replied:
Or is this nowadays pulled in / defined by mocha plug-in?

barmac replied:
This is not defined by mocha plugin ([mocha](https://github.com/lo1tuma/eslint-plugin-mocha/blob/main/index.js#L126), [globals](https://github.com/sindresorhus/globals/blob/HEAD/globals.json#L2075)). I believe we should re-add it.

barmac replied:
We could also consider adding `sinon` to the globals as we have done in diagram-js: https://github.com/bpmn-io/diagram-js/commit/7de8f7732b2f42f079b6c0c2b2dcb526fbfd78d0
It's commonly used in bpmn.io ecosystem.

barmac replied:
So mocha, chai, and sinon.

barmac replied:
Chai can be added via globals: https://github.com/sindresorhus/globals/blob/HEAD/globals.json#L1217

nikku replied:
We don't need chai, but `expect` only, this is our pattern.

nikku reviewed (commented):
Could you explain how the "errors" / "no errors" is supposed to work? I see that you commented out all errors. Would such a test fail if the line did not yield the particular error?

barmac commented on test/package.json:
```diff
@@ -2,11 +2,19 @@
   "scripts": {
-    "test": "eslint ."
+    "test": "npm-run-all test:*",
+    "test:browser": "eslint browser --config browser/eslint.config.js --report-unused-disable-directives",
```
Smart.

barmac commented on test/mocha/browser/errors.spec.js:
```diff
@@ -0,0 +1,38 @@
+import { expect } from 'chai';
+
```
Related to https://github.com/bpmn-io/eslint-plugin-bpmn-io/pull/20/files#r1791477214

```suggestion
```

barmac commented:
> Could you explain how the "errors" / "no errors" is supposed to work? I see that you commented out all errors. Would such a test fail if the line did not yield the particular error?

As far as I understand, the errors specs contain overrides which are reported if eslint does not use them, cf. https://github.com/bpmn-io/eslint-plugin-bpmn-io/pull/20#pullrequestreview-2354008236

So if I have a snippet below, the script will report unused directive:

```javascript
const life = 42; // eslint-disable-line no-unused-vars
console.log(life);
```

nikku commented:
I spent quite some time trying to dig how eslint@9 handles overrides / globals. My goal was to write a [test case](https://github.com/bpmn-io/eslint-plugin-bpmn-io/pull/20#pullrequestreview-2353848936) that would pass with `expect` being a global, but fail if we don't declare it.

I propose that we sync (knowledge exchange) on the topic later this week @philippfromme. Seems to be of inherent complexity :see_no_evil:

nikku committed:
chore: organize <test> in a child workspace

Allows us to get rid of some hacks.

nikku commented:
Ok, I went on to drastically simplify this PR. From what it looks like https://github.com/bpmn-io/eslint-plugin-bpmn-io/pull/20/commits/9be702a00be8c83043c241b82acee86752375c3c is a requirement to properly (later on) be able to have [simple, per branch configurations](https://github.com/bpmn-io/eslint-plugin-bpmn-io/pull/20/commits/857e53b7281f9a52c7a1ec2c31618fc6f5c8b25c#diff-95cb9363e4dfaa058fce17610d8cb4f0d5e3a0cdcc2546b9295799a2006bc536).

@barmac @philippfromme Please review.

nikku commented:
What I'd be interested in is whether the "full ES6" style of configuring things has a negative impact in different places (i.e. when consuming in Desktop Modeler).

:arrow_right: I guess not, because we can provide the configuration in an `mjs` file, [as you propose](https://github.com/camunda/camunda-modeler/commit/209303d45b66bb429718cdc9dd58937c0176bfad#diff-9601a8f6c734c2001be34a2361f76946d19a39a709b5e8c624a2a5a0aade05f2).

barmac commented on configs/jsx.js:
```diff
@@ -1,13 +1,8 @@
 import reactPlugin from 'eslint-plugin-react';
 
 export default [
+  reactPlugin.configs.flat.recommended,
```
I think this is a good change. We can always override the recommended setup if we need to.

barmac commented:
This pull request includes significant changes to modernize the codebase by transitioning from CommonJS to ES Modules, updating ESLint configurations, and improving the test setup. The most important changes include modifying the ESLint configurations, updating the CI workflow, and restructuring the test files.

### ESLint Configuration Updates:
* [`configs/browser.js`](diffhunk://#diff-f4cdf060029099e4d941a60b8c83c981a3e0d94f43f8a88800790f28c1ff21a3L1-R22): Refactored to use ES Modules and updated the configuration to better align with modern JavaScript standards.
* [`configs/jsx.js`](diffhunk://#diff-8bee03180d0f106cc80ac2dd45c1250699ce3398f5d153ec031463eda47ec278L1-R24): Transitioned to ES Modules and updated the React-specific linting rules.
* [`configs/mocha.js`](diffhunk://#diff-9858773e9c2dcd842531c4945db33ae156458e6dd0f1b88f2f8c9b5028c83abdL1-R15): Updated to use ES Modules and improved the Mocha-specific linting rules.
* [`configs/node.js`](diffhunk://#diff-bbe609df5412f30b30e38202d1911467de29f8f83971aa967e051fc59b4f44d5L1-R14): Refactored to use ES Modules and updated the Node.js-specific linting rules.
* [`configs/recommended.js`](diffhunk://#diff-836ec658cdd636bac9d283925cbdef7754bbb3ad21532eee581c08746671b2bcL1-R45): Transitioned to ES Modules and enhanced the recommended linting rules.

### CI Workflow:
* [`.github/workflows/CI.yml`](diffhunk://#diff-3ab46ee209a127470fce3c2cf106b1a1dbadbb929a4b5b13656a4bc4ce19c0b8L10-R10): Updated the `eslint-version` matrix to use version 9 exclusively.

### Test Setup:
* [`test/browser/eslint.config.js`](diffhunk://#diff-a48e2329741b871ae39d94ae4a65b6bc47a7237dbafafebe0312f99609d37fc2R1-R12): Added a new ESLint configuration for browser-specific tests.
* [`test/browser/fixtures/errors.js`](diffhunk://#diff-b317d89bba9539193b56862b7e31555fed8880c387b9bea88591750bd58b8f63R1-R50): Added new test fixtures to validate ESLint rules.
* [`test/custom/eslint.config.js`](diffhunk://#diff-8f3ccdb80f06c3257fee498503cdd2c9b353815c2742b1e67cfef4db9fd36b1dR1-R13): Created a custom ESLint configuration for testing purposes.
* [`test/custom/fixtures/errors.js`](diffhunk://#diff-dde83d550ad684afdfe8315a2d680934b0c2ede52a9d1f7a5a3b0e1884a92d4fR1-R49): Added new test fixtures for custom ESLint rules.

### Module Conversion:
* [`index.js`](diffhunk://#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346L1-R7): Converted the main entry point to use ES Modules and updated the export syntax.
* [`package.json`](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L5-R16): Updated to specify the module type and adjusted the exports field accordingly. [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L5-R16) [[2]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L23-R43)

### Cleanup:
* Removed obsolete or empty configuration files (`test/.babelrc`, `test/.eslintrc`, `test/.gitignore`, `test/browser/.eslintrc`, `test/custom/.eslintrc`). [[1]](diffhunk://#diff-e63cce22913753b3d2a2b8af5116b1fbbb369b930744e3be0f6fa91c5fb206e8L1-L3) [[2]](diffhunk://#diff-481e628c4ad8bb249c0e964a98e5416f9a00b7cbc013e1ef9f0314e249688203L1-L3) [[3]](diffhunk://#diff-3fa66f55d79a0b5bb5a5c4a76d2388236984f16369d4e41761fd9e3743f67a0fL1) [[4]](diffhunk://#diff-50ac47e891f4e4b9a95f38508cd8e55bc650c42e7344313070c4a9c73ad2f7fbL1-L8) [[5]](diffhunk://#diff-ae64b63e4e96cad200e4270cc7207a33205bab63022328b7503c65a8581bcc07L1-L9)

barmac commented:
LGTM 🚀

barmac commented on package.json:
```diff
@@ -2,14 +2,18 @@
   "version": "1.0.1",
   "description": "Shared lint rules for bpmn.io projects",
-  "main": "index.js",
+  "type": "module",
```
Let's add a breaking change to commit.

philippfromme replied:
👍🏻

philippfromme replied:
Updated the commit message to:

### feat: migrate to ESLint 9 and flat configs

BREAKING CHANGE: This plugin is now an ES module compatible with ESLint 9.

- Node.js version that supports ES modules is now required
- ESLint 9.0.0 is now required
- Import statements must be used instead of require() in ES module projects
- dynamic import() or the @eslint/compat package must be used in CommonJS projects

Closes https://github.com/bpmn-io/eslint-plugin-bpmn-io/issues/19
Related to https://github.com/bpmn-io/internal-docs/issues/1042

philippfromme replied:
@barmac @nikku Can you have a look?

barmac replied:
Add: In a commonjs project, the eslint configuration needs to use `.mjs` extension to be recognized by Node.

barmac replied:
But looks good overall

barmac replied:
I see Nico is making some changes too

nikku replied:
Cf. https://github.com/bpmn-io/eslint-plugin-bpmn-io/pull/20/commits/b1362f3643d679043dfa53f1d3fdd41122c2edc5, https://github.com/bpmn-io/eslint-plugin-bpmn-io/pull/20/commits/dd859631f875fd1e5b089a67cf96298deb408ca6, https://github.com/bpmn-io/eslint-plugin-bpmn-io/pull/20/commits/e14ca88628df70613aed03b9227e55f38ed92c6c.

I think we're good to merge this.

nikku committed:
feat!: turn into ES module, migrate to ESLint 9 and flat configs

  Closes #19

  Related to https://github.com/bpmn-io/internal-docs/issues/1042

  BREAKING CHANGE: require eslint@9 (drop support for <= eslint@8)
  BREAKING CHANGE: use from ESM code-base, this module is ESM only
  BREAKING CHANGE: consume as ES module

feat!: work off react:recommended and mocha:recommended rules

nikku committed:
test: migrate to flat config + eslint@9

nikku committed:
test: move specs to fixtures, introduce error specs

nikku committed:
docs(README): document modern way of working

nikku committed:
chore(CHANGELOG): update to v2.0.0

barmac commented on .github/workflows/CI.yml:
```diff
@@ -7,7 +7,7 @@ jobs:
         os: [ macos-latest, ubuntu-latest, windows-latest ]
         node-version: [ 18, 20, 22 ]
-        eslint-version: [ 8, '' ]
+        eslint-version: [ 9 ]
```
If we change it as it is now, eslint@9 will be [reinstalled](https://github.com/bpmn-io/eslint-plugin-bpmn-io/pull/20/files#diff-3ab46ee209a127470fce3c2cf106b1a1dbadbb929a4b5b13656a4bc4ce19c0b8R33) each time CI executes. The version list is necessary for backwards compatibility testing. Alternatively, we could remove the "Install eslint" step and the commented line as we don't test backwards compatibility anymore.

```suggestion
        eslint-version: [ '' ]
```

nikku replied:
:+1:

barmac commented:
One more thing [required]

barmac commented:
OK done via https://github.com/bpmn-io/eslint-plugin-bpmn-io/pull/20/commits/decea2781c75c720eb17c7ed5fee353a5af0d460. Check out the comment if you want to know details.

I'd say we can squash it now.

nikku commented:
All cleaned up and ready to merge :rocket:

barmac commented:
Congratz @philippfromme and @nikku for a successful migration

philippfromme commented:
Thanks for the help, guys! 💪🏻
